AX: Accessibility objects should live as long their DOM node, not as long as their renderer, which is destroyed and recreated frequently, causing unnecessary churn for ATs

twilco opened the pull request:
#### 7bee0b21b7d3b68d6ef5ffd2219e3db3faf677a8
<pre>
AX: Accessibility objects should live as long their DOM node, not as long as their renderer, which is destroyed and recreated frequently, causing unnecessary churn for ATs
<a href="https://bugs.webkit.org/show_bug.cgi?id=296959">https://bugs.webkit.org/show_bug.cgi?id=296959</a>
<a href="https://rdar.apple.com/157598062">rdar://157598062</a>

Reviewed by Andres Gonzalez.

Prior to this commit, accessibility objects only lived as long as their renderer did. This is problematic because renderers
are destroyed and recreated constantly, as they exist to serve rendering. This can cause extreme instability for assistive
technologies, who get a reference to an object and try to use it, and find it has been destroyed, when really the renderer
was refreshed but the DOM node is still there. I have traced multiple bugs over the years to this behavior, e.g.
VoiceOver infinitely looping over content on reddit.com, allowing AT text searches to auto-expand details elements, etc.

With this commit, DOM nodes become the &quot;primary key&quot; for any accessibility object that has one. This required a slew
of changes to work around the fact that we can no longer rely on constantly throwing away and recreating objects to have
fresh (non-stale) state:

  - AXObjectCache::m_deferredRendererChangedList added to handle necessary updates to objects who lose or gain a renderer.

  - AXObjectCache::m_deferredRecomputeActiveSummaryList added to re-compute summary / details role and ignored state
    when the active summary changes (we used to rely on render tree updates destroying and recreating objects to update the accessibility tree)

  - Detect when menu elements become ignored (e.g. because of display:none) and post a notification

  - Update cached WebAccessibilityObjectWrapperIOS::m_isAccessibilityElement when an object changed ignored state

  - Add a case in AccessibilityRenderObject::nextSibling() to use the DOM to get the next sibling when we compute a next-sibling
    that is ourselves. This happens because continuations can result in multiple renderers sharing the same node, so if
    renderer-A computes renderer-B as its next sibling, and either are part of the same continuation, they&apos;ll compute to
    the same accessibility object, causing an infinite loop. New test slotted-url-continuation-hang.html added reproducing this bug.

  - Fix an existing bug where nodes / renderers deleted from iframes never result in a call to AXObjectCache::remove(Node/RenderObject&amp;).
    This happened because it seems sometimes iframe-documents can lose their m_frame, meaning they cannot access a Page, and thus
    cannot access an AXObjectCache, meaning the node / renderer removal never gets seen by the AXObjectCache. This again always
    existed, but manifested as a layout test failure with this PR, which lead me to the discovery. Resolve it by adding a new
    Document::m_topAXObjectCache field, which can stay alive even after the document&apos;s frame is inexplicably lost.

Beyond object-stability for ATs, this will likely have efficiency benefits for the isolated tree. When AXIsolatedTree::updateChildren
detects that children have changed (e.g. because a renderer subtree got blown away and recreated, meaning brand new objects, wrappers,
and AXIDs), we do a lot of work. That won&apos;t happen anymore.

* LayoutTests/accessibility/display-contents/table-dynamic.html:
* LayoutTests/accessibility/menu-list-crash2-expected.txt:
* LayoutTests/accessibility/menu-list-crash2.html:
* LayoutTests/accessibility/slider-with-lost-renderer-expected.txt:
* LayoutTests/accessibility/slider-with-lost-renderer.html:
* LayoutTests/accessibility/slotted-url-continuation-hang-expected.txt: Added.
* LayoutTests/accessibility/slotted-url-continuation-hang.html: Added.
* LayoutTests/platform/ios/accessibility/menu-list-crash2-expected.txt:
* LayoutTests/platform/ios/accessibility/slider-with-lost-renderer-expected.txt:
* LayoutTests/resources/accessibility-helper.js:
(dumpAXSearchTraversal):
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::linkedObjects const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::composedParentIgnoringDocumentFragments):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::cacheAndInitializeWrapper):
(WebCore::AXObjectCache::getOrCreate):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::onRendererCreated):
(WebCore::AXObjectCache::onDetailsSummarySlotChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::onPaint const):
(WebCore::layoutTableCellRole):
(WebCore::AXObjectCache::get const): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityList.cpp:
(WebCore::AccessibilityList::computeIsIgnored const):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
(WebCore::AccessibilityNodeObject::create): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::extraDebugInfo const):
(WebCore::AccessibilityObject::postMenuClosedNotificationIfNecessary const):
(WebCore::AccessibilityObject::detachRemoteParts):
(WebCore::AccessibilityObject::isContinuation const):
(WebCore::AccessibilityObject::isIgnoredWithoutCache const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::markPlatformWrapperIgnoredStateDirty const):
(WebCore::AccessibilityObject::extraDebugInfo const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::create):
(WebCore::AccessibilityRenderObject::nextSibling const):
(WebCore::AccessibilityRenderObject::selectedText const):
(WebCore::AccessibilityRenderObject::selectedTextRange const):
(WebCore::AccessibilityRenderObject::insertionPointLineNumber const):
(WebCore::AccessibilityRenderObject::setSelectedTextRange):
(WebCore::AccessibilityRenderObject::visiblePositionForIndex const):
(WebCore::AccessibilityRenderObject::indexForVisiblePosition const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
(WebCore::AccessibilityRenderObject::setRenderer):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::extraDebugInfo const):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::AccessibilityObject::markPlatformWrapperIgnoredStateDirty const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper determineIsAccessibilityElement]):
(-[WebAccessibilityObjectWrapper _clearCachedIsAccessibilityElementState]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::processQueuedNodeUpdates):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::clearAXObjectCache):
(WebCore::Document::existingAXObjectCacheSlow const):
(WebCore::Document::axObjectCache const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::DetailsSlotAssignment::hostChildElementDidChange):

Canonical link: <a href="https://commits.webkit.org/298426@main">https://commits.webkit.org/298426@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0a54bd3fefad10df61670830a3f6f1a78d10871

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115437 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25630 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121522 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66010 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35788 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43699 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87718 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118385 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28543 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103615 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68110 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27701 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21737 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65178 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97932 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21850 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124687 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42372 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31745 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96490 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42739 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99803 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96276 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24497 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41505 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19361 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38274 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42252 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47812 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41755 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45082 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43472 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1606,9 +1606,9 @@ void AXIsolatedTree::processQueuedNodeUpdates()
 
     if (m_mostRecentlyPaintedTextIsDirty) {
+        m_mostRecentlyPaintedTextIsDirty = false;
         Locker lock { m_changeLogLock };
         m_pendingMostRecentlyPaintedText = cache->mostRecentlyPaintedText();
-        m_mostRecentlyPaintedTextIsDirty = false;
```
Drive-by change — `m_mostRecentlyPaintedTextIsDirty` is only read and written by the main-thread, so there is no need to be holding the lock while we set it.

minorninth commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1212,6 +1202,16 @@ void AXObjectCache::remove(RenderObject& renderer)
     AXTRACE(makeString("AXObjectCache::remove RenderObject* 0x"_s, hex(reinterpret_cast<uintptr_t>(this))));
     remove(m_renderObjectMapping.takeOptional(renderer));
+
+    if (RefPtr node = renderer.node()) {
```
If it has a node, then the remove(m_renderObjectMapping.takeOptional(renderer)) above should be a no-op, right? Could we have an assertion that if it has a node, that it's not in m_renderObjectMapping?

twilco replied:
Agreed, a fix will be included for this in the next commit!

minorninth commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1300,46 +1300,17 @@ void AXObjectCache::handleTextChanged(AccessibilityObject* object)
-            // only serves to repair relationships for replaced objects, which text nodes cannot
-            // possibly be part of (because they are not elements).
-        }
+    if (RefPtr existingObject = dynamicDowncast<AccessibilityRenderObject>(get(node))) {
```
It looks like this handles the case where we already have an object associated with the node. What if we don't know about this object yet? Is this a safe time to call getOrCreate(node)? Are we just assuming we'd get a different callback about the node?

twilco replied:
> Is this a safe time to call getOrCreate(node)?

This is not a safe time to call `getOrCreate`, since this function is called mid-layout.

> What if we don't know about this object yet?

That's perfectly fine — when the object is `getOrCreate`d later (at a safe time), we'll grab the renderer off the node then. This function only needs to handle renderer changes for existing objects.

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -179,13 +179,28 @@ String AccessibilityObject::debugDescriptionInternal(bool verbose, std::optional
-    if (detachmentType == AccessibilityDetachmentType::ElementDestroyed && role() == AccessibilityRole::Menu) {
-        if (auto* cache = axObjectCache())
-            cache->postNotification(nullptr, cache->document(), AXNotification::MenuClosed);
+    if (isContinuation())
```
Should this be on AccessibilityNodeObject? I feel like (for now at least) AccessibilityObject is supposed to be for common code that doesn't assume you have a node or a renderer, right?

twilco replied:
Removed this method entirely. It was not necessary for this change.

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -247,6 +249,7 @@ class AccessibilityObject : public AXCoreObject {
     Node* node() const override { return nullptr; }
     RenderObject* renderer() const override { return nullptr; }
     RenderObject* rendererOrNearestAncestor() const;
+    bool isContinuation() const;
```
I feel like somewhere we need a comment explaining what this is and why it matters. The next person who joins the team won't know what a continuation is.

What if we called it nodeMayHaveMultipleRenderers() instead, where "is continuation" is the implementation detail, or maybe that could be used for the YouTube embed?

twilco replied:
Removed this method entirely. It was not necessary for this change.

minorninth commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -449,10 +454,20 @@ AccessibilityObject* AccessibilityRenderObject::nextSibling() const
     if (nextSibling->node() && nextSibling->node() == m_renderer->node()) {
-        if (RefPtr nextObject = cache->getOrCreate(*nextSibling))
+        if (RefPtr nextObject = cache->getOrCreate(*nextSibling)) {
+            if (nextObject.get() == this) {
```
Just checking, does this handle the case where there are multiple siblings associated with the same node that aren't adjacent? It feels like this is just handling the case where it's the next sibling, but what about the next next?

twilco replied:
It doesn't — I guess I'd want some motivating example before changing any code to accommodate this. If such an example does exist, I'd rather solve it by changing Cocoa ports to always walk the DOM rather than walk the render tree, instead of trying to implement some complicated cycle detection here.

minorninth commented on Source/WebCore/accessibility/AccessibilityRenderObject.h:
```diff
@@ -55,8 +55,18 @@ class VisibleSelection;
+    static Ref<AccessibilityRenderObject> create(AXID, Node&, AXObjectCache&);
     virtual ~AccessibilityRenderObject();
 
+    // Returns true if the renderer changed.
```
Should we just name this setRendererAndReturnIfChanged?

Or maybe setRendererIfNeeded would be a more common pattern and would imply the return value?

minorninth commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.h:
```diff
@@ -56,6 +56,7 @@ static NSString * const UIAccessibilityTextualContextSourceCode = @"UIAccessibil
 - (BOOL)accessibilityCanFuzzyHitTest;
 
 - (BOOL)isAccessibilityElement;
+- (void)_clearCachedIsAccessibilityElementState;
```
Maybe internal methods like this should be in a different section and not grouped with the public accessibility methods like isAccessibilityElement and accessibilityLabel?

minorninth commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -1068,6 +1071,11 @@ - (BOOL)isAccessibilityElement
 
+- (void)_clearCachedIsAccessibilityElementState
+{
+    m_isAccessibilityElement = -1;
```
I know this was pre-existing, but could we just make this a tristate enum or at least have some constants? I hate magic numbers and sooner or later one of us will have to debug it and hunt around figuring out what -1 means

minorninth reviewed (approved):
A few thoughts, but overall looks really clean and comprehensive. Nice work!

jesxilin commented on LayoutTests/accessibility/menu-list-crash2.html:
```diff
@@ -19,31 +20,32 @@
-        setTimeout(async function() {
-            menulist.style.display = "none";
-            gc();
+    // var includeIgnoredInCoreAXTreeEnabled = accessibilityController.platformName === "mac";
```
I might not have enough context but was this comment left for a reason?

twilco replied:
Oops, no 😅 removed

minorninth reviewed (approved):
lgtm!

AndresGonzalezApple commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1519,6 +1519,9 @@ AXCoreObject::AccessibilityChildrenVector AXCoreObject::linkedObjects() const
 
+    linkedObjects.removeAllMatching([] (const auto& object) {
+        return object->isIgnored();
+    });
```
Haven't we moved the isIgnored filter to the platform layer yet? I was under the impression that we did.

twilco replied:
We haven't.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -892,8 +885,14 @@ Ref<AccessibilityNodeObject> AXObjectCache::createFromNode(Node& node)
+            return AccessibilityLabel::create(AXID::generate(), *element, *this);
     }
-    return AccessibilityNodeObject::create(AXID::generate(), node, *this);
+    return AccessibilityRenderObject::create(AXID::generate(), node, *this);
```
Why do we create an AccessibilityRenderObject now instead of an AccessibilityNodeObject?

twilco replied:
Now that accessibility objects live for the lifetime of the node, we need to handle the case where a DOM node naturally gains or loses a renderer without the DOM node itself changing (e.g. toggling between `display:contents` and `display:block`). `AccessibilityNodeObject` does not have an `m_renderer` field, so if an instance of this class later gained a renderer, we'd have nowhere to put it — hence creating an `AccessibilityRenderObject` here.

I think the long-term vision is for `AccessibilityRenderObject` to be squished into `AccessibilityNodeObject` (or some other name), with that class representing objects with a renderer, a node, or both.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1211,7 +1212,18 @@ void AXObjectCache::remove(std::optional<AXID> axID)
 {
     AXTRACE(makeString("AXObjectCache::remove RenderObject* 0x"_s, hex(reinterpret_cast<uintptr_t>(this))));
-    remove(m_renderObjectMapping.takeOptional(renderer));
+    ASSERT(!renderer.node() || !m_renderObjectMapping.contains(renderer));
```
Shouldn't this be ASSERT(renderer.node() || m_renderObjectMapping.contains(renderer)) ?

twilco replied:
The intent of the assert is to verify that if something has a node, it is not present in the m_renderObjectMapping (it should instead be in m_nodeObjectMapping), which I believe is accomplished by what is written. We assert that either the object doesn't have a node, or if it does have a node, that it is not contained in m_renderObjectMapping. Does that seem accurate with your reading of this?

AndresGonzalezApple replied:
I think the condition you are trying to check is (renderer.node() && !m_renderObjectMapping.contains(renderer)) || (!renderer.node() && m_renderObjectMapping.contains(renderer)), which is a logical difference or exclusive XOR, I.e.:
(renderer.node() ^ m_renderObjectMapping.contains(renderer)), but I may be still misreading.

twilco replied:
I think the latter condition you propose:

`(!renderer.node() && m_renderObjectMapping.contains(renderer))`

Makes the assumption that every node-less render object that gets created and removed has had an accessibility object created for it, which probably isn't true, and doesn't need to be true — it's OK for this function to be a no-op for renderers that have no corresponding accessibility object.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4742,19 +4733,20 @@ void AXObjectCache::performDeferredCacheUpdate(ForceLayout forceLayout)
+        object.recomputeIsIgnored();
+    }
+    m_deferredRecomputeActiveSummaryList.clear();
 
```
Do we need this separate list for summaries since we are doing exactly the same thing we are doing to the list above? E.g., can we have all objects including the summaries in the same list?

twilco replied:
The semantics of these lists are really different — the fact that they need to do the same thing to update the accessibility tree is coincidence. I could very much see these behaviors eventually diverging too (e.g. if we decide use `m_deferredRendererChangedList` to drive the mechanism for destroying and recreating objects when we detect a subclass change). So I feel like keeping them separate feels right, but let me know if you have another idea for this.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -5231,7 +5223,13 @@ void AXObjectCache::onPaint(const RenderObject& renderer, IntRect&& paintRect) c
+    if (RefPtr node = renderer.node())
+        axID = m_nodeObjectMapping.getOptional(*node);
+    if (!axID)
+        axID = m_renderObjectMapping.getOptional(const_cast<RenderObject&>(renderer));
```
Should we add a function to get the AXID that does the above? E.g., getAXID(RednerObject&).

twilco replied:
Good idea! Fixed.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -5231,7 +5223,13 @@ void AXObjectCache::onPaint(const RenderObject& renderer, IntRect&& paintRect) c
+        axID = m_nodeObjectMapping.getOptional(*node);
+    if (!axID)
+        axID = m_renderObjectMapping.getOptional(const_cast<RenderObject&>(renderer));
+    m_geometryManager->cacheRect(axID, WTFMove(paintRect));
```
Not that we were checking before, but is it possible that we add a null AXID here? Does it matter?

twilco replied:
Hmm, yeah I think a null-check here is a good idea. Will add.

twilco replied:
Fixed!

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -5374,6 +5372,13 @@ bool isRenderHidden(const RenderStyle* style)
+    // This return value is what will be used if AccessibilityTableCell determines
+    // the cell should not be treated as a cell (e.g. because it is a layout table).
+    return AccessibilityRole::TextGroup;
+}
```
This doesn't belong here, I think it should be in AXCoreObject or one of its subclasses (AX cell ?).

twilco replied:
It's a free-standing helper function, not associated with any class really. I followed the pattern used with our other helper functions — define in `AXCoreObject.h`, implement in `AXCoreObject.cpp`. Ultimately I want to use it from both `AccessibilityNodeObject.cpp` and `AccessibilityRenderObject.cpp`, so I could move it to `AccessibilityNodeObject.h` and `AccessibilityNodeObject.cpp`?

twilco replied:
Fixed this!

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -350,12 +350,14 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
+        if (RefPtr object = get(renderer.node()))
+            return object.get();
         auto axID = m_renderObjectMapping.getOptional(renderer);
         return axID ? m_objects.get(*axID) : nullptr;
```
This could be written more elegantly if we add the getAXID(RenderObject&) function suggested above.
Since we are touching this, should we move the inline implementation outside the class declaration?

twilco replied:
Added the `getAXID(RenderObject&)`, which does clean this up nicely!

Regarding the inline implementations — I think you may be right, but that's predicated on having a different header called `FooInlines.h`, e.g. `NodeInlines.h` so that the compiler still can inline the function even though it's not directly in `Foo.h`. We don't have such a thing for `AXObjectCache`, so maybe we should take that up as a refactoring in a later PR, as we definitely do want these functions to be inlined for performance.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -374,7 +376,13 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
+        auto nodeID = m_nodeObjectMapping.get(node);
+        return nodeID ? m_objects.get(*nodeID) : nullptr;
+    }
 
```
Dito about inline implementation.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -936,9 +938,8 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
-    // attached a renderer, but then want to replace it with a new AX object after the renderer has been attached.
-    HashSet<AXID> m_deferredReplacedObjects;
+    WeakHashSet<AccessibilityObject> m_deferredRendererChangedList;
+    WeakHashSet<AccessibilityObject> m_deferredRecomputeActiveSummaryList;
```
Can we merge these two?

twilco replied:
Touched on this in a previous comment — I don't think so.

AndresGonzalezApple commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -1079,6 +1080,8 @@ bool isRenderHidden(const RenderStyle&);
 // Only checks CSS visibility properties.
 bool isVisibilityHidden(const RenderStyle&);
 
+AccessibilityRole layoutTableCellRole();
```
Move to AXCoreObject or subclass.

AndresGonzalezApple commented on Source/WebCore/accessibility/AccessibilityList.cpp:
```diff
@@ -66,6 +66,9 @@ Ref<AccessibilityList> AccessibilityList::create(AXID axID, Node& node, AXObject
+    if (!m_renderer)
+        return AccessibilityNodeObject::computeIsIgnored();
+
     return isIgnoredByDefault();
```
Maybe more concise as ? : statement.

twilco replied:
Fixed!

AndresGonzalezApple commented on Source/WebCore/accessibility/AccessibilityNodeObject.h:
```diff
@@ -41,7 +41,6 @@ class Node;
 
 class AccessibilityNodeObject : public AccessibilityObject {
 public:
-    static Ref<AccessibilityNodeObject> create(AXID, Node&, AXObjectCache&);
```
So, we are enforcing not to create AccessibilityNodeObject directly? But instead on of the subclasses? That change seems independent of the main change in this patch.

twilco replied:
This is a required part of this overall change, as described in another comment.

AndresGonzalezApple replied:
Right, but it could be done as a prequel change, especially with the long term goal that you mentioned above to get rid of the AXNodeObject class. But since it is already here, it is fine.

AndresGonzalezApple commented on Source/WebCore/accessibility/AccessibilityRenderObject.h:
```diff
@@ -55,8 +55,18 @@ class VisibleSelection;
+        if (m_renderer == renderer)
+            return false;
+        m_renderer = renderer;
+        return true;
```
Thought that WebKit style doesn't want inline implementations in the class declaration, but my memory is thin here.

AndresGonzalezApple commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.h:
```diff
@@ -43,9 +43,15 @@ static NSString * const UIAccessibilityTextualContextSourceCode = @"UIAccessibil
+    No,
+    Yes,
+    Unknown
+};
```
This one can be written in one line, I think WebKit style permits.

twilco replied:
Fixed!

twilco committed:
AX: Accessibility objects should live as long their DOM node, not as long as their renderer, which is destroyed and recreated frequently, causing unnecessary churn for ATs
https://bugs.webkit.org/show_bug.cgi?id=296959
rdar://157598062

Reviewed by Andres Gonzalez.

Prior to this commit, accessibility objects only lived as long as their renderer did. This is problematic because renderers
are destroyed and recreated constantly, as they exist to serve rendering. This can cause extreme instability for assistive
technologies, who get a reference to an object and try to use it, and find it has been destroyed, when really the renderer
was refreshed but the DOM node is still there. I have traced multiple bugs over the years to this behavior, e.g.
VoiceOver infinitely looping over content on reddit.com, allowing AT text searches to auto-expand details elements, etc.

With this commit, DOM nodes become the "primary key" for any accessibility object that has one. This required a slew
of changes to work around the fact that we can no longer rely on constantly throwing away and recreating objects to have
fresh (non-stale) state:

  - AXObjectCache::m_deferredRendererChangedList added to handle necessary updates to objects who lose or gain a renderer.

  - AXObjectCache::m_deferredRecomputeActiveSummaryList added to re-compute summary / details role and ignored state
    when the active summary changes (we used to rely on render tree updates destroying and recreating objects to update the accessibility tree)

  - Detect when menu elements become ignored (e.g. because of display:none) and post a notification

  - Update cached WebAccessibilityObjectWrapperIOS::m_isAccessibilityElement when an object changed ignored state

  - Add a case in AccessibilityRenderObject::nextSibling() to use the DOM to get the next sibling when we compute a next-sibling
    that is ourselves. This happens because continuations can result in multiple renderers sharing the same node, so if
    renderer-A computes renderer-B as its next sibling, and either are part of the same continuation, they'll compute to
    the same accessibility object, causing an infinite loop. New test slotted-url-continuation-hang.html added reproducing this bug.

  - Fix an existing bug where nodes / renderers deleted from iframes never result in a call to AXObjectCache::remove(Node/RenderObject&).
    This happened because it seems sometimes iframe-documents can lose their m_frame, meaning they cannot access a Page, and thus
    cannot access an AXObjectCache, meaning the node / renderer removal never gets seen by the AXObjectCache. This again always
    existed, but manifested as a layout test failure with this PR, which lead me to the discovery. Resolve it by adding a new
    Document::m_topAXObjectCache field, which can stay alive even after the document's frame is inexplicably lost.

Beyond object-stability for ATs, this will likely have efficiency benefits for the isolated tree. When AXIsolatedTree::updateChildren
detects that children have changed (e.g. because a renderer subtree got blown away and recreated, meaning brand new objects, wrappers,
and AXIDs), we do a lot of work. That won't happen anymore.

* LayoutTests/accessibility/display-contents/table-dynamic.html:
* LayoutTests/accessibility/menu-list-crash2-expected.txt:
* LayoutTests/accessibility/menu-list-crash2.html:
* LayoutTests/accessibility/slider-with-lost-renderer-expected.txt:
* LayoutTests/accessibility/slider-with-lost-renderer.html:
* LayoutTests/accessibility/slotted-url-continuation-hang-expected.txt: Added.
* LayoutTests/accessibility/slotted-url-continuation-hang.html: Added.
* LayoutTests/platform/ios/accessibility/menu-list-crash2-expected.txt:
* LayoutTests/platform/ios/accessibility/slider-with-lost-renderer-expected.txt:
* LayoutTests/resources/accessibility-helper.js:
(dumpAXSearchTraversal):
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::linkedObjects const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::composedParentIgnoringDocumentFragments):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::cacheAndInitializeWrapper):
(WebCore::AXObjectCache::getOrCreate):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::onRendererCreated):
(WebCore::AXObjectCache::onDetailsSummarySlotChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::onPaint const):
(WebCore::layoutTableCellRole):
(WebCore::AXObjectCache::get const): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityList.cpp:
(WebCore::AccessibilityList::computeIsIgnored const):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
(WebCore::AccessibilityNodeObject::create): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::extraDebugInfo const):
(WebCore::AccessibilityObject::postMenuClosedNotificationIfNecessary const):
(WebCore::AccessibilityObject::detachRemoteParts):
(WebCore::AccessibilityObject::isContinuation const):
(WebCore::AccessibilityObject::isIgnoredWithoutCache const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::markPlatformWrapperIgnoredStateDirty const):
(WebCore::AccessibilityObject::extraDebugInfo const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::create):
(WebCore::AccessibilityRenderObject::nextSibling const):
(WebCore::AccessibilityRenderObject::selectedText const):
(WebCore::AccessibilityRenderObject::selectedTextRange const):
(WebCore::AccessibilityRenderObject::insertionPointLineNumber const):
(WebCore::AccessibilityRenderObject::setSelectedTextRange):
(WebCore::AccessibilityRenderObject::visiblePositionForIndex const):
(WebCore::AccessibilityRenderObject::indexForVisiblePosition const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
(WebCore::AccessibilityRenderObject::setRenderer):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::extraDebugInfo const):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::AccessibilityObject::markPlatformWrapperIgnoredStateDirty const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper determineIsAccessibilityElement]):
(-[WebAccessibilityObjectWrapper _clearCachedIsAccessibilityElementState]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::processQueuedNodeUpdates):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::clearAXObjectCache):
(WebCore::Document::existingAXObjectCacheSlow const):
(WebCore::Document::axObjectCache const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::DetailsSlotAssignment::hostChildElementDidChange):

Canonical link: https://commits.webkit.org/298426@main
